File: primes.py
#!/usr/bin/env python2

from sys import stdout, argv, exit

RATE = 5000

def percentProgress(step, total):
    return 100.0 * (float(step) / float(total))

def countPrimes(maxVal):
    nPrimes = 0
    primes = []

    for i in xrange(2, maxVal):
        prime = True

        for p in primes:
            # Exit loop if not prime or not going to find factor
            if (p * p > i):
                break

            # Exit loop if proven not prime
            if ((i % p) == 0):
                prime = False
                break

        if prime:
            primes.append(i)
            nPrimes += 1
            if (nPrimes < 100):
                stdout.write("{:3}, ".format(i))
                if not nPrimes % 5:
                    stdout.write("\n")
            elif (nPrimes == 100):
                stdout.write("{:3}\n".format(i))

            if ((nPrimes % RATE) == 0):
                stdout.write("\033[GFound {:12d} at {:7.3f}%...".format(
                    nPrimes, percentProgress(i, maxVal)
                ))
                stdout.flush()

    if nPrimes < 100 and primes % 5:
        stdout.write("\n")

    stdout.write("\033[G\033[KFound {:12d} primes.\n".format(nPrimes))

if __name__ == "__main__":
    if len(argv) != 2:
        print "Usage: {} max_prime".format(argv[0])
        exit(1)

    maxVal = int(argv[1])

    if (maxVal <= 0):
        print "max_val must be a positive integer"
        exit(1)

    countPrimes(maxVal)

File: primes.rs
use std::io::{stdout, Write};
use std::str::FromStr;
use std::{env, process};

const RATE: u32 = 50000;

fn main() {
    let mut args = env::args();

    if args.len() != 2 {
        println!("Usage: {} max_prime", args.nth(0).unwrap());
        process::exit(1);
    }

    let max = u32::from_str(args.nth(1).unwrap().as_str()).unwrap_or(0);

    if max <= 0 {
        println!("max_prime must bu a positive integer");
        process::exit(1);
    }

    count_primes(max);
}

fn percent_progress(step: u32, total: u32) -> f64 {
    100.0f64 * (step as f64 / total as f64)
}

fn count_primes(max: u32) {
    let mut n_primes = 0u32;
    let mut primes: Vec<u32> = Vec::new();
    let mut sout = stdout();

    for i in 2..(max + 1) {
        let mut prime = true;

        for j in &primes {
            if (j * j) > i {
                break;
            }

            if (i % j) == 0 {
                prime = false;
                break;
            }
        }

        if prime {
            primes.push(i);
            n_primes += 1;

            if n_primes < 100 {
                print!("{:3}, ", i);
                if (n_primes % 5) == 0 {
                    println!("");
                }
            } else {
                println!("{:3}", i);
            }

            if (n_primes % RATE) == 0 {
                print!(
                    "\x1B[GFound {:12} at {:7.3}%...",
                    n_primes,
                    percent_progress(i, max)
                );
                sout.flush().unwrap();
            }
        }
    }

    if n_primes < 100 && (n_primes % 5) != 0 {
        println!("");
    }

    println!("\x1B[G\x1B[KFound {:12} primes.", n_primes);
}
